src/colmap/estimators/bundle_adjustment.cc

```diff
@@ -810,6 +813,257 @@ void RigBundleAdjuster::ParameterizeCameraRigs(Reconstruction* reconstruction) {
+      LOG(INFO) << "Sim3 alignment error w.r.t. prior position:";
+      LOG(INFO) << "  - rmse:   " << std::sqrt(Mean(verr2_wrt_prior)) << " m";
+      LOG(INFO) << "  - median: " << std::sqrt(Median(verr2_wrt_prior)) << " m";
+
+    } else {
+      LOG(WARNING) << "Sim3 alignment w.r.t. prior position failed!";
+    }
+  }
+
+  // Put the centroid of the reconstruction to (0.,0.,0.)
+  // to avoid any numerical instability
+  Sim3d sim_to_center;
+  for (const auto& image_id : config_.Images()) {
+    sim_to_center.translation -=
+        reconstruction->Image(image_id).ProjectionCenter();
+  }
+  sim_to_center.translation /= config_.NumImages();
+
+  reconstruction->Transform(sim_to_center);
+
+  // Transform the priors according to the centroid projection
+  for (const auto image_id : config_.Images()) {
+    auto& image = reconstruction->Image(image_id);
+    if (image.HasPosePrior()) {
+      image.WorldFromCamPrior().position =
+          sim_to_center * image.WorldFromCamPrior().position;
+    }
+  }
+
+  SetUpProblem(
+      reconstruction, loss_function_.get(), prior_loss_function_.get());
+
+  if (problem_->NumResiduals() == 0) {
+    return false;
+  }
+
+  ceres::Solver::Options solver_options =
+      SetUpSolverOptions(*problem_, options_.solver_options);
+
+  ceres::Solve(solver_options, problem_.get(), &summary_);
+
+  // Unproject centroid from (0.,0.,0.)
+  const Sim3d sim_from_center = Inverse(sim_to_center);
+  reconstruction->Transform(sim_from_center);
+  for (const auto image_id : config_.Images()) {
+    auto& image = reconstruction->Image(image_id);
+    if (image.HasPosePrior()) {
+      image.WorldFromCamPrior().position =
+          sim_from_center * image.WorldFromCamPrior().position;
+    }
+  }
+
+  if (options_.print_summary || VLOG_IS_ON(1)) {
+    PrintSolverSummary(summary_, "Position Prior Bundle adjustment report");
+  }
+
+  return true;
+}
+
+void PositionPriorBundleAdjuster::SetUpProblem(
+    Reconstruction* reconstruction,
+    ceres::LossFunction* loss_function,
+    ceres::LossFunction* prior_loss_function) {
+  THROW_CHECK_NOTNULL(reconstruction);
+
+  // Initialize an empty problem
+  ceres::Problem::Options problem_options;
+  problem_options.loss_function_ownership = ceres::DO_NOT_TAKE_OWNERSHIP;
+  problem_ = std::make_shared<ceres::Problem>(problem_options);
+
+  // Set up problem
+  // Warning: AddPointsToProblem assumes that AddImageToProblem is called first.
+  // Do not change order of instructions!
+  for (const image_t image_id : config_.Images()) {
+    AddImageToProblem(
+        image_id, reconstruction, loss_function, prior_loss_function);
+  }
+  for (const auto point3D_id : config_.VariablePoints()) {
+    AddPointToProblem(point3D_id, reconstruction, loss_function);
+  }
+  for (const auto point3D_id : config_.ConstantPoints()) {
+    AddPointToProblem(point3D_id, reconstruction, loss_function);
+  }
+
+  ParameterizeCameras(reconstruction);
+  ParameterizePoints(reconstruction);
+}
```

Comment:
Mostly identical to `BundleAdjuster::SetUpProblem`, can we reuse it?

Reply:
I have just improved the code to better re-use functions from the BundleAdjuster class.

Reply:
I think that `AddPosePriorToProblem` should be called from `SetUpProblem`. You could simply call the parent `BundleAdjuster::SetUpProblem` at the beginning of `PositionPriorBundleAdjuster::SetUpProblem`.

Reply:
:+1: I have also added a few more factorization which makes the PosePriorBundleAdjuster looks great IMO.